### PR TITLE
Image block: Add support for "more" dropdown for additional tools in Write mode

### LINKED ISSUE #66455: Add support for "more" dropdown for additional tools in Write mode
When writing in contentOnly mode, selecting an image block will render tools for adding a link, adding alt text, and adding title text (if there is title text present on the block already. This proposal here is oriented towards keeping the "Write" tooling simple and intuitive, leveraging the existing "More" down caret employed within text blocks, for additional controls. 

This way we can continue flexing content blocks within Write mode, without introducing further complexity. 

### Visuals

Let's start with the Image block, moving the "Alternative text" control inside this more menu, as well as surfacing the "Title" control, instead of having it only render when a title exists previously. 

Image

Image

### Requirements 

- Provide a mechanism for write mode block controls to be defined within a more menu. 
- When a block control in the more menu is engaged, it should be active in the toolbar and complete its associated action. For example, the "Alternative text" control engages the popover with the control for modifying the alternative text. 
- Modify the image block's write mode block controls as a basis for this effort.

Related to #66054 and #66116. 

Edited by @richtabor

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/66455

>When writing in contentOnly mode, selecting an image block will render tools for adding a link, adding alt text, and adding title text (if there is title text present on the block already. This proposal here is oriented towards keeping the "Write" tooling simple and intuitive, leveraging the existing "More" down caret employed within text blocks, for additional controls.

>This way we can continue flexing content blocks within Write mode, without introducing further complexity.

## Testing Instructions
1. Insert an image and go to `write` mode
2. Tinker with the new 'more menu' by adding/removing values, saving, changing selected block etc..

## Screenshots or screencast 

https://github.com/user-attachments/assets/3e6d340b-e3be-424d-95cd-76814835b55a

### CODE REVIEW

File: packages/block-library/src/image/image.js
Comment (ntsekouras) [inline packages/block-library/src/image/image.js]: We need to find a proper description here.

File: packages/block-library/src/image/image.js
Comment (ntsekouras) [inline packages/block-library/src/image/image.js]: If we have both the `more` menu and the active toolbar buttons in the same BlockControls group, they can be rendered after the menu. I'm not sure if we have so fine control over the order of fills in this slot, but I think not..

**Comment by jasmussen** [comment]
Hey, thanks for tackling this. But I'm seeing this:

Screenshot 2024-10-31 at 09 47 03

I think we might need some input from @richtabor as he had some input on #66455 as well, and I want to make sure I'm not missing nuance. But the way I see it, the items in the "More" dropdown should never add _block toolbar options_, and the items Title and Alternative text should never show up in the block toolbar. The options could potentially open modals, or otherwise, where you can input the values you need. But the purpose is to simplify the block toolbar, not to add conditionals.

**Comment by t-hamano** [comment]
I posted an idea here: https://github.com/WordPress/gutenberg/issues/66455#issuecomment-2449097788

**Comment by jasmussen** [comment]
> I posted an idea here: https://github.com/WordPress/gutenberg/issues/66455#issuecomment-2449097788

Apologies for missing that. I love that idea.

Though one bit of feedback, I would add a vertical separator between "Replace" and the "More" dropdown. Conceptually the "More" dropdown is grouped with the "additional" tools like Link, just like it would be for a Paragraph block, so there needs to be that vertical black line separator still, if there's no link button.

**Comment by ntsekouras** [comment]
>I think we might need some input from @richtabor as he had some input on https://github.com/WordPress/gutenberg/issues/66455 as well, and I want to make sure I'm not missing nuance. But the way I see it, the items in the "More" dropdown should never add block toolbar options, and the items Title and Alternative text should never show up in the block toolbar.

That what I got from the issue screenshots.. I'll see to update with the suggestions.

File: packages/block-library/src/image/image.js
Comment (mcsf) [reply packages/block-library/src/image/image.js]: I'd remove the ellipsis, but other than that this matches what we do `FormatToolbar`, which has been worked on by a11y experts.

We _could_ try to be more descriptive (i.e. find something accurate and concise like "display more content-editing controls", etc.), but in the end couldn't that verbosity get in the way of the user? The sooner they'll expand the dropdown menu, the sooner they can get a true sense of what hides in there.

(Precise and verbose descriptions can also get in the way of translation.)

**Review by mcsf — APPROVED** [review]
Looking good!

~Since this is still marked as a draft, is there anything left to do? Or is it a matter of design approval?~
**Edit:** my approval only considers the implementation

**Comment by jasmussen** [comment]
Appreciate the review but let's not merge this, I think Aki is on to something with his idea though!

**Comment by ntsekouras** [comment]
<del> I think we should add a Modal for editing the values as it's more consistent with what we do in other places. I'll update soon. </del>

I followed the suggestion by @t-hamano eventually. It's similar to some RichText controls.

**Comment by jasmussen** [comment]
Seeing this:

test

Nice work, toolbar-wise, this hits the mark in terms of appearance, functionality, etc.

The now open question is the "toggled states" inside the dropdown. It's not clear to me those are useful at all. What purpose do they serve, can can we just remove those states, so that the menu items are just menu items?

**Comment by ntsekouras** [comment]
>The now open question is the "toggled states" inside the dropdown. It's not clear to me those are useful at all. What purpose do they serve, can can we just remove those states, so that the menu items are just menu items?

I removed them.

**Comment by jasmussen** [comment]
Happy to ship this from my end 👍 👍

**Review by mcsf — APPROVED** [review]
Looks great, go for it!

File: packages/block-library/src/image/image.js
Comment (talldan) [inline packages/block-library/src/image/image.js:150]: As these open another popover they should probably have `aria-haspopup` - https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup.

When pressing escape from the alt or title popovers, focus doesn't go back to these buttons, which is a bit awkward and it's not really clear what happened, was the change successful or not? A user has to reopen it to find out if the changes stuck which then takes two presses.

**Comment by talldan** [comment]
When an image is used in a synced pattern as an override, it looks a bit unusual having a more menu at the start of the toolbar on its own:
Screenshot 2024-11-07 at 4 45 06 pm

The menu does lose a bit of context when there are no other options in the same group, it took me a while to realize what was happening. 🤔 

It will get a bit better in the future if block bindings/pattern overrides can support the link option, but I'm not sure where that support is at right now.

File: packages/block-library/src/image/image.js
Comment (ntsekouras) [reply packages/block-library/src/image/image.js:150]: I added the aria prop [here](https://github.com/WordPress/gutenberg/pull/66815).

Regarding the design/UX questions I'd also defer to @jasmussen or @jameskoster . It's fairly new so it would need a bit time to see if we have some different feedback from users.

While focus doesn't go back to the button that triggered it, there is no focus loss though. The use case you describe requires two clicks but I'm not sure if someone who writes something and clicks escape goes back to a dialog to see if it worked - it should have worked, no?

Finally I followed some RichText tools approach (which was the same as the suggested design to close the main menu and open a dialog). That doesn't mean that it might be wrong, but if it is we should update that too.

Comment (jasmussen) [reply packages/block-library/src/image/image.js:150]: I've no strong opinions on the behavior, full-modal or otherwise, and would support whichever solves the problem.

**Comment by jasmussen** [comment]
> The menu does lose a bit of context when there are no other options in the same group, it took me a while to realize what was happening. 🤔

It's not clear how much control we have over these contentOnly toolbars, but specifically from the perspective of that Image block, I would think of Upload being in the same group as the "More" control. I.e.—rough mockup—like this:

```
[ image | upload more | replace | reset ]
```

**Comment by talldan** [comment]
> It's not clear how much control we have over these contentOnly toolbars, but specifically from the perspective of that Image block, I would think of Upload being in the same group as the "More" control. I.e.—rough mockup—like this:

@jasmussen I had a go at moving the upload button. It is possible, but not straightforward because of the way slot fills work (same issue as https://github.com/WordPress/gutenberg/issues/15641). Though a lot of the time these images might not have upload button either if they're from the media library:
Screenshot 2024-11-08 at 1 00 01 pm

Let me know what you think and if it's worth following up on.

**Comment by jasmussen** [comment]
If the code isn't too ugly or problematic, and doesn't take too much of your time, I think it would be a good move 🙏

**Pull request merged**
